llmclient/llms.py

```diff
@@ -586,76 +549,82 @@ async def check_rate_limit(self, token_count: float, **kwargs) -> None:
 
-    @rate_limited
-    async def acomplete_iter(  # type: ignore[override]
-        self, prompt: str
-    ) -> AsyncIterable[Chunk]:
-        completion = await track_costs_iter(self.router.atext_completion)(
-            model=self.name,
-            prompt=prompt,
-            stream=True,
-            stream_options={"include_usage": True},
-        )
-        async for chunk in completion:
-            yield Chunk(
-                text=chunk.choices[0].text, prompt_tokens=0, completion_tokens=0
-            )
-        if hasattr(chunk, "usage") and hasattr(chunk.usage, "prompt_tokens"):
-            yield Chunk(
-                text=chunk.choices[0].text, prompt_tokens=0, completion_tokens=0
+        # We are not streaming here, so we can cast to list[litellm.utils.Choices]
+        choices = cast(list[litellm.utils.Choices], completions.choices)
+        for completion in choices:
+            if completion.finish_reason == "tool_calls" or getattr(
+                completion.message, "tool_calls", None
+            ):
+                serialized_message = completion.message.model_dump()
+                serialized_message["tool_calls"] = (
+                    serialized_message.get("tool_calls") or []
+                )
+                output_messages: list[Message | ToolRequestMessage] = [
+                    ToolRequestMessage(**serialized_message)
+                ]
+            else:
+                output_messages = [Message(**completion.message.model_dump())]
+            results.append(
+                LLMResult(
+                    model=self.name,
+                    text=completion.message.content,
+                    prompt=messages,
+                    messages=output_messages,
+                    logprob=sum_logprobs(completion),
+                    prompt_count=completions.usage.prompt_tokens,  # type: ignore[attr-defined]
+                    completion_count=completions.usage.completion_tokens,  # type: ignore[attr-defined]
+                    system_fingerprint=completions.system_fingerprint,
+                )
             )
+        return results
 
     @rate_limited
-    async def achat(self, messages: list[Message]) -> Chunk:  # type: ignore[override]
-        prompts = [m.model_dump(by_alias=True) for m in messages if m.content]
-        response = await track_costs(self.router.acompletion)(self.name, prompts)
-        return Chunk(
-            text=cast(litellm.Choices, response.choices[0]).message.content,
-            prompt_tokens=response.usage.prompt_tokens,  # type: ignore[attr-defined]
-            completion_tokens=response.usage.completion_tokens,  # type: ignore[attr-defined]
-        )
-
-    @rate_limited
-    async def achat_iter(  # type: ignore[override]
-        self, messages: list[Message]
-    ) -> AsyncIterable[Chunk]:
+    async def acompletion_iter(  # type: ignore[override]
+        self, messages: list[Message], **kwargs
+    ) -> AsyncGenerator[LLMResult]:
```

Comment:
```suggestion
    ) -> AsyncIterator[LLMResult]:
```

Base class has `AsyncIterator[LLMResult]`, why does this have `AsyncGenerator[LLMResult]`?

Reply:
Oh, nice catch. Both were `AsyncIterator` before. But while changing types in `rate_limited` I noticed that it is just `yield`ing the `LLMResults`. So I think this is actually an `AsyncGenerator`. Forgot to adjust the base class. What do you think? Is it correct that `acompletion_iter` should return an `AsyncGenerator`?